### DISCUSSION TITLE
Blazor dependency injection in helper class

### DISCUSSION THREAD

**Discussion by user72356** [dotnet/aspnetcore#26357] [General] [answered]
I'm building a helper class for some of my components, that wraps an external JS library. This class doesn't need to be a component itself, but it needs to be instantiated by each component needing to use the JS lib.

I can't figure out how to gain access to the IJSRuntime service within the helper class. Constructor DI doesn't seem to be working, nor is using the [Inject] parameter.

I don't want to manually pass IJSRuntime to the class because I don't want the components to be aware of internal implementation details such as reliance on IJSRuntime.

I tried inheriting my helper class from ComponentBase, constructor DI still not working, but @inject and [Inject] work ONLY if I place the element on the razor page. If I keep the element off the page and I tried to instantiate it with new, I'm not getting DI.

**Comment by SteveSandersonMS** [comment, marked as answer]
> I don't want to manually pass IJSRuntime to the class because I don't want the components to be aware of internal implementation details such as reliance on IJSRuntime.

I'm afraid you have to, because otherwise how would it know which `IJSRuntime` to use? With dependency injection in general, there are multiple "scopes" representing different partitions of services. For example with Blazor Server, there are multiple concurrent users, and each one has a separate `IJSRuntime` instance. To obtain a service instance, you need to request it from a particular scope. Components can do this because they are attached to a particular renderer that holds a reference to the scope.

In general, the DI system doesn't hook into .NET's `new` keyword. So there's no way for it to automatically populate properties or constructor parameters on things you create directly with `new`.

What you probably want to do is register your helper as a DI service itself. Then components that want to use the helper can `@inject MyHelperType Helper`. This means they will be constructed using the DI system, which in turn will take care of supplying any constructor parameters that are other DI services such as `IJSRuntime`.
